// @subsquid/util-internal-validation/lib/primitives.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BYTES = exports.SMALL_QTY = exports.QTY = exports.NAT = exports.INT = exports.STRING = void 0;
const error_1 = require("./error");
exports.STRING = {
    cast(value) {
        if (typeof value == 'string') {
            return value;
        }
        else {
            return new error_1.ValidationFailure(value, '{value} is not a string');
        }
    },
    validate(value) {
        if (typeof value == 'string')
            return;
        return new error_1.ValidationFailure(value, '{value} is not a string');
    },
    phantom() {
        return '';
    }
};
/**
 * Safe integer
 */
exports.INT = {
    cast(value) {
        if (isInteger(value)) {
            return value;
        }
        else {
            return new error_1.ValidationFailure(value, '{value} is not an integer');
        }
    },
    validate(value) {
        if (isInteger(value))
            return;
        return new error_1.ValidationFailure(value, '{value} is not an integer');
    },
    phantom() {
        return 0;
    }
};
function isInteger(value) {
    return typeof value == 'number' && Number.isSafeInteger(value);
}
/**
 * Safe integer greater or equal to 0
 */
exports.NAT = {
    cast(value) {
        if (isInteger(value) && value >= 0) {
            return value;
        }
        else {
            return new error_1.ValidationFailure(value, '{value} is not a safe natural number');
        }
    },
    validate(value) {
        if (isInteger(value) && value >= 0)
            return;
        return new error_1.ValidationFailure(value, '{value} is not a safe natural number');
    },
    phantom() {
        return 0;
    }
};
function isBytes(value) {
    return typeof value == 'string' && /^0x[0-9a-f]*$/.test(value);
}
/**
 * Hex encoded natural number of an arbitrary size
 */
exports.QTY = {
    cast(value) {
        if (isBytes(value)) {
            return BigInt(value);
        }
        else {
            return new error_1.ValidationFailure(value, `{value} is not a hex encoded natural number`);
        }
    },
    validate(value) {
        if (isBytes(value))
            return;
        return new error_1.ValidationFailure(value, `{value} is not a hex encoded natural number`);
    },
    phantom() {
        return '0x0';
    }
};
/**
 * Hex encoded safe natural number
 */
exports.SMALL_QTY = {
    cast(value) {
        if (isBytes(value)) {
            let val = parseInt(value);
            if (Number.isSafeInteger(val)) {
                return val;
            }
            else {
                return new error_1.ValidationFailure(value, `{value} is not a safe integer`);
            }
        }
        else {
            return new error_1.ValidationFailure(value, `{value} is not a hex encoded natural number`);
        }
    },
    validate(value) {
        let i = this.cast(value);
        if (i instanceof error_1.ValidationFailure)
            return i;
    },
    phantom() {
        return '0x0';
    }
};
/**
 * Hex encoded binary string
 */
exports.BYTES = {
    cast(value) {
        return this.validate(value) || value;
    },
    validate(value) {
        if (isBytes(value))
            return;
        return new error_1.ValidationFailure(value, `{value} is not a hex encoded binary string`);
    },
    phantom() {
        return '0x';
    }
};
